File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";
import { z } from "zod";

// ==== Type Definitions, feel free to add or modify ==========================
const requiredItem = z.object({
  name: z.string(),
  quantity: z.number().int().min(0),
});

const recipe = z.object({
  type: z.literal("recipe"),
  name: z.string(),
  requiredItems: z.array(requiredItem).superRefine((items, ctx) => {
    const seen = new Set<string>();
    for (let i = 0; i < items.length; i++) {
      const lowerName = items[i].name.toLowerCase();
      if (seen.has(lowerName)) {
        ctx.addIssue({
          code: "custom",
          message: `Duplicate required item: ${items[i].name}`,
        });
      }
      seen.add(lowerName);
    }
  }),
});

const ingredient = z.object({
  type: z.literal("ingredient"),
  name: z.string(),
  cookTime: z.number().int().min(0), 
});

const cookbookEntrySchema = z.discriminatedUnion("type", [
  recipe,
  ingredient,
]);

// Auto infer types
type CookbookEntry = z.infer<typeof cookbookEntrySchema>;
type RequiredItem = z.infer<typeof requiredItem>;

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: Record<string, CookbookEntry> = {};
const cache: Record<string, { cookTime: number; ingredients: Record<string, number> }> = {};

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  if (!recipeName) return null;

  let cleaned_name = recipeName.replace(/[-_]/g, " ").replace(/[^a-zA-Z\s]/g, "");
  const words = cleaned_name.split(/\s+/).filter((word) => word.length > 0);
  
  const final_name = words
    .map((word) => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(" ");

  return final_name.length > 0 ? final_name : null;
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const data = req.body;

  if (!data || typeof data !== "object") {
    return res.status(400).send("Invalid JSON");
  }

  const parsedData = cookbookEntrySchema.safeParse(data);
  if (!parsedData.success) {
    return res.status(400).json({ error: parsedData.error.message });
  }

  const entry = parsedData.data;
  if (cookbook[entry.name]) {
    return res.status(400).json({ error: `Entry with name '${entry.name}' already exists` });
  }

  cookbook[entry.name] = entry;
  return res.status(200).send({});
});


// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name

const get_ingredients = (itemName: string): { cookTime: number; ingredients: Record<string, number> } => {
  if (cache[itemName]) return cache[itemName];

  const item = cookbook[itemName];
  if (!item) {
    throw new Error(`Item '${itemName}' not found in cookbook`);
  }

  if (item.type === "ingredient") {
    const result = { cookTime: item.cookTime, ingredients: { [itemName]: 1 } };
    cache[itemName] = result;
    return result;
  }

  let totalTime = 0;
  const ingredientCounts: Record<string, number> = {};

  for (const reqItem of item.requiredItems) {
    const reqResult = get_ingredients(reqItem.name);

    totalTime += reqResult.cookTime * reqItem.quantity;

    for (const [ingName, ingCount] of Object.entries(reqResult.ingredients)) {
      ingredientCounts[ingName] = (ingredientCounts[ingName] || 0) + ingCount * reqItem.quantity;
    }
  }

  const result = { cookTime: totalTime, ingredients: ingredientCounts };
  cache[itemName] = result;
  return result;
};

app.get("/summary", (req:Request, res:Request) => {
  const { name } = req.query;
  if (!name || typeof name !== "string" || !cookbook[name]) {
    return res.status(400).json({ error: "Invalid 'name' query parameter" });
  }

  if (cookbook[name].type !== "recipe") {
    return res.status(400).json({ error: "Requested entry is not a recipe" });
  }

  try {
    const { cookTime, ingredients } = get_ingredients(name);

    const summary_ingredients: RequiredItem[] = Object.entries(ingredients).map(
      ([ingName, quantity]) => ({
        name: ingName,
        quantity: quantity,
      })
    );

    return res.status(200).json({
      name,
      cookTime,
      ingredients: summary_ingredients,
    });
  } catch (error: any) {
    return res.status(400).json({ error: error.message });
  }
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
import re
from typing import Annotated, List, Literal, Union

from flask import Flask, jsonify, request
from pydantic import BaseModel, Field, TypeAdapter, ValidationError, field_validator


# ==== Type Definitions, feel free to add or modify ===========================
class RequiredItem(BaseModel):
    name: str
    quantity: int


class Recipe(BaseModel):
    type: Literal["recipe"]
    name: str
    requiredItems: List[RequiredItem]

    @field_validator("requiredItems")
    @classmethod
    def validate_required_items(cls, items):
        seen = set()
        for item in items:
            if (item.name.lower()) in seen:
                raise ValueError(f"Duplicate required item: {item.name}")
            seen.add(item.name.lower())
        return items


class Ingredient(BaseModel):
    type: Literal["ingredient"]
    name: str
    cookTime: Annotated[int, Field(ge=0)]


Cookbook = Annotated[Union[Recipe, Ingredient], Field(discriminator="type")]
CookbookChecker = TypeAdapter(Cookbook)


class RecipeSummary(BaseModel):
    name: str
    cookTime: int
    ingredients: List[RequiredItem]


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}


# Task 1 helper (don't touch)
@app.route("/parse", methods=["POST"])
def parse():
    data = request.get_json()
    recipe_name = data.get("input", "")
    parsed_name = parse_handwriting(recipe_name)
    if parsed_name is None:
        return "Invalid recipe name", 400
    return jsonify({"msg": parsed_name}), 200


# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that
def parse_handwriting(recipeName: str) -> Union[str | None]:
    if not recipeName:
        return None

    cleaned_name = re.sub(r"[-_]", " ", recipeName)
    cleaned_name = re.sub(r"[^a-zA-Z\s]", "", cleaned_name)
    words = cleaned_name.split()
    final_name = " ".join(word.capitalize() for word in words)

    if len(final_name) > 0:
        return final_name

    return None


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route("/entry", methods=["POST"])
def create_entry():
    data = request.get_json()

    if not data:
        return "Invalid JSON", 400

    try:
        entry = CookbookChecker.validate_python(data)
    except ValidationError as e:
        return jsonify({"error": e.errors()}), 400

    if entry.name in cookbook:
        return jsonify({"error": f"Entry with name '{entry.name}' already exists"}), 400

    cookbook[entry.name] = entry
    return {}, 200


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name

cache = {}


@app.route("/summary", methods=["GET"])
def summary():
    name = request.args.get("name")
    if not name or name not in cookbook:
        return jsonify({"error": "Recipe not found"}), 400

    if cookbook[name].type != "recipe":
        return jsonify({"error": "Requested entry is not a recipe"}), 400

    def get_ingredients(item_name: str):
        if item_name in cache:
            return cache[item_name]

        item = cookbook.get(item_name)
        if not item:
            raise ValueError(f"Item '{item_name}' not found in cookbook")

        if item.type == "ingredient":
            result = (item.cookTime, {item_name: 1})
            cache[item_name] = result
            return result

        total_time = 0
        ingredient_counts = {}
        for req_item in item.requiredItems:
            req_time, req_ingredients = get_ingredients(req_item.name)
            total_time += req_time * req_item.quantity
            for ing_name, ing_count in req_ingredients.items():
                ingredient_counts[ing_name] = (
                    ingredient_counts.get(ing_name, 0) + ing_count * req_item.quantity
                )

        result = (total_time, ingredient_counts)
        cache[item_name] = result
        return result

    try:
        total_cook_time, ingredients = get_ingredients(name)
    except ValueError as e:
        return jsonify({"error": str(e)}), 400

    summary_ingredients = [
        RequiredItem(name=ing_name, quantity=ing_count)
        for ing_name, ing_count in ingredients.items()
    ]

    summary = RecipeSummary(
        name=name, cookTime=total_cook_time, ingredients=summary_ingredients
    )

    return summary.model_dump(), 200


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == "__main__":
    app.run(debug=True, port=8080)
